docs(react): writing tests with testing-library (#2861)
ionic-team/ionic-docs: docs/react/testing/introduction.md

---
sidebar_label: Introduction
title: Ionic React Testing Introduction
description: Learn how to test an Ionic React application. This document provides an overview of how to test an application built with @ionic/react.
---

# Testing Ionic React

> This document provides an overview of how to test an application built with `@ionic/react`. It covers the basics of testing with React, as well as the specific tools and libraries used in `@ionic/react`.

Reviewer: These libraries aren't used in `@ionic/react`, so this wording is misleading. Maybe we could say something like the following instead:
```suggestion
This document provides an overview of how to test an application built with `@ionic/react`. It covers the basics of testing with React, as well as tools and libraries developers can use to test their applications.
```

## Introduction

Testing is an important part of the development process, and it helps to ensure that an application is working as intended. In `@ionic/react`, testing is done using a combination of tools and libraries, including Vitest, Jest, React Testing Library, Playwright or Cypress.

## Types of Tests

There are two types of tests that can be written:

> **Unit Tests**: Unit tests are used to test individual functions and components in isolation. They can be written using tools such as [Jest](https://jestjs.io) and [React Testing Library](https://testing-library.com) or [Vitest](https://vitest.dev).

Reviewer: This wording is a bit confusing. Does this mean to group Jest+Testing Library as a single "unit" of what someone would use to test?

> **Unit Tests**: Unit tests are used to test individual functions and components in isolation. They can be written using tools such as [Jest](https://jestjs.io) and [React Testing Library](https://testing-library.com) or [Vitest](https://vitest.dev).

Reviewer: Why do we mention Vitest if we never show anything with it in this guide?

---
sidebar_label: Introduction
title: Ionic React Testing Introduction
description: Learn how to test an Ionic React application. This document provides an overview of how to test an application built with @ionic/react.
---

# Testing Ionic React

This document provides an overview of how to test an application built with `@ionic/react`. It covers the basics of testing with React, as well as the specific tools and libraries used in `@ionic/react`.

## Introduction

Testing is an important part of the development process, and it helps to ensure that an application is working as intended. In `@ionic/react`, testing is done using a combination of tools and libraries, including Vitest, Jest, React Testing Library, Playwright or Cypress.

## Types of Tests

There are two types of tests that can be written:

> **Unit Tests**: Unit tests are used to test individual functions and components in isolation. They are typically written using [Jest](https://jestjs.io) and [React Testing Library](https://testing-library.com) or [Vitest](https://vitest.dev).

Reviewer: ```suggestion
**Unit Tests**: Unit tests are used to test individual functions and components in isolation. They can be written using tools such as [Jest](https://jestjs.io) and [React Testing Library](https://testing-library.com) or [Vitest](https://vitest.dev).
```

Reviewer: The current phrasing reads that those are the only tools devs can use with Ionic React.

> **Integration Tests**: Integration tests are used to test how different components work together. They are typically written using [Cypress](https://www.cypress.io) or [Playwright](https://playwright.dev).

Reviewer: ```suggestion
**Integration Tests**: Integration tests are used to test how different components work together. They can be written using tools such as [Cypress](https://www.cypress.io) or [Playwright](https://playwright.dev).
```

> **Integration Tests**: Integration tests are used to test how different components work together. They are typically written using [Cypress](https://www.cypress.io) or [Playwright](https://playwright.dev).

Author: Yes, eventually my goal is we build out sections/recommendations for setting up integration tests for your Ionic React apps (and the other frameworks as well, potentially a shared guide). 

For this stage, it is more about informing there is two different "types of tests" and we will have documentation on setting up unit testing for now.

## Tools and Libraries

### Unit Testing

[**Jest**](https://jestjs.io): Jest is a popular testing framework for JavaScript applications. It is included with `react-scripts` and is used to run unit tests.

[**Vitest**](https://vitest.dev): Vitest is a modern testing framework powered by Vite. It is compatible with Jest and is used to run unit tests.

> [**React Testing Library**](https://testing-library.com): React Testing Library is a set of utilities that make it easier to test React components. It's used to interact with components and test their behavior.

Reviewer: It looks like https://github.com/ionic-team/ionic-docs/pull/2861/commits/6c01f2842260b3e924dd21a713412e69fa560fcd  got reverted at some point

### Integration Testing

[**Cypress**](https://www.cypress.io): Cypress is a modern testing tool that's used to run end-to-end tests. It allows developers to test the entire application, including the user interface, in a simulated environment.

[**Playwright**](https://playwright.dev): Playwright is a browser automation library that's used to run end-to-end tests. It allows developers to test the entire application, including the user interface, in a simulated environment.

## Additional Resources

> - [Testing Ionic React Apps with Jest and React Testing Library](https://ionicframework.com/blog/testing-ionic-react-apps-with-jest-and-react-testing-library/)

Reviewer: Should we still link to this? If we aren't recommending @ionic/react-test-utils this might be confusing.

Author: Good call, I think with the discovery you did in the `requestAnimationFrame` mocking, we can nix this section.
